server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -696,7 +697,15 @@ private void raisePhaseFailure(SearchPhaseExecutionException exception) {
      * @see #onShardResult(SearchPhaseResult, SearchShardIterator)
      */
     final void onPhaseDone() {  // as a tribute to @kimchy aka. finishHim()
-        executeNextPhase(this, getNextPhase(results, this));
+        final SearchPhase nextPhase = getNextPhase(results, this);
+        if (request instanceof PipelinedRequest && nextPhase != null) {
+            // From src files the next phase is never null, but from tests this is a possibility. Hence, making sure that
+            // tests pass, we need to do null check on next phase.
+            if (nextPhase != null) {
+                ((PipelinedRequest) request).transformSearchPhase(results, this, this.getName(), nextPhase.getName());
```

Comment:
The problem is results is marked as final variable. So we cannot assign it. 

https://github.com/opensearch-project/OpenSearch/blob/213c7c4a3c0f49d225d30f97784b77891eb2a6c9/server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java#L98

I think the better solution will be the SearchPhaseResultsProcessor mutate the object rather than changing the reference. I can make that till downstream, @reta please let me know if that removes this concern.

Reply:
The `SearchPhaseResults` looks like immutable holder so we should not make it mutable (becomes very difficult to reason about) but we need to find out the way to compose the results or, do we really need `transformSearchPhase` to mutate anything on  `SearchPhaseResults` side?

Reply:
I did some changes in the revision, please let me know does that resolve the issue.